Mathlib/Algebra/Lie/Weights.lean

```diff
@@ -166,64 +133,72 @@ protected theorem weight_vector_multiplication (M₁ : Type w₁) (M₂ : Type w
   · rw [(hf_comm.pow_pow i j).eq, LinearMap.mul_apply, LinearMap.pow_map_zero_of_le hi hf₁,
       LinearMap.map_zero]
   · rw [LinearMap.mul_apply, LinearMap.pow_map_zero_of_le hj hf₂, LinearMap.map_zero]
-#align lie_module.weight_vector_multiplication LieModule.weight_vector_multiplication
-
-variable {L M}
 
-theorem lie_mem_preWeightSpace_of_mem_preWeightSpace {χ₁ χ₂ : L → R} {x : L} {m : M}
-    (hx : x ∈ preWeightSpace L χ₁) (hm : m ∈ preWeightSpace M χ₂) :
-    ⁅x, m⁆ ∈ preWeightSpace M (χ₁ + χ₂) := by
-  apply LieModule.weight_vector_multiplication L L M M (toModuleHom R L M) χ₁ χ₂
+lemma lie_mem_maxGenEigenspace_toEndomorphism
+    {χ₁ χ₂ : R} {x y : L} {m : M} (hy : y ∈ 𝕎 L χ₁ x) (hm : m ∈ 𝕎 M χ₂ x) :
+    ⁅y, m⁆ ∈ 𝕎 M (χ₁ + χ₂) x := by
+  apply LieModule.weight_vector_multiplication L M M (toModuleHom R L M) χ₁ χ₂
   simp only [LieModuleHom.coe_toLinearMap, Function.comp_apply, LinearMap.coe_comp,
     TensorProduct.mapIncl, LinearMap.mem_range]
-  use ⟨x, hx⟩ ⊗ₜ ⟨m, hm⟩
+  use ⟨y, hy⟩ ⊗ₜ ⟨m, hm⟩
   simp only [Submodule.subtype_apply, toModuleHom_apply, TensorProduct.map_tmul]
-#align lie_module.lie_mem_pre_weight_space_of_mem_pre_weight_space LieModule.lie_mem_preWeightSpace_of_mem_preWeightSpace
 
 variable (M)
 
-/-- If a Lie algebra is nilpotent, then pre-weight spaces are Lie submodules. -/
-def weightSpace [LieAlgebra.IsNilpotent R L] (χ : L → R) : LieSubmodule R L M :=
-  { preWeightSpace M χ with
-    lie_mem := fun hm => by
-      dsimp only
+/-- If `M` is a representation of a nilpotent Lie algebra `L`, `χ` is a scalar, and `x : L`, then
+`weightSpaceOf M χ x` is the maximal generalized `χ`-eigenspace of the action of `x` on `M`.
+
+It is a Lie submodule because `L` is nilpotent. -/
+def weightSpaceOf [LieAlgebra.IsNilpotent R L] (χ : R) (x : L) : LieSubmodule R L M :=
+  { 𝕎 M χ x with
+    lie_mem := by
+      intro y m hm
+      simp only [AddSubsemigroup.mem_carrier, AddSubmonoid.mem_toSubsemigroup,
+        Submodule.mem_toAddSubmonoid] at hm ⊢
       rw [← zero_add χ]
-      refine lie_mem_preWeightSpace_of_mem_preWeightSpace ?_ hm
-      suffices preWeightSpace L (0 : L → R) = ⊤ by simp only [this, Submodule.mem_top]
-      exact LieAlgebra.iInf_max_gen_zero_eigenspace_eq_top_of_nilpotent R L }
-#align lie_module.weight_space LieModule.weightSpace
+      exact lie_mem_maxGenEigenspace_toEndomorphism (by simp) hm }
+
```

Comment:
Would it be possible to change the meaning of `𝕎 M χ x`, so that it now denotes a weight space? Or do you think that's not a good idea?

Reply:
I'd be very reluctant to have the same notation mean different things in this way.

I claim the notation for `(toEndomorphism R L M x).maximalGeneralizedEigenspace χ` is useful because it's such a mouthful and we will never care about it (in statements anyway) once we define `weightSpaceOf M χ x`.

However I think `weightSpace M χ` is perfectly fine and not in need of notation.

Of course I'm open to suggestions but as I say, I'd be reluctant to overload notation depending on context.

Reply:
Fair enough, makes a lot of sense